### ISSUE TITLE
Issue 005: `src` 非変更で Coarse train に `trend_prior_loss` を追加する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#290]
# Issue 005: `src` 非変更で Coarse train に `trend_prior_loss` を追加する

## 背景

更新版 repo には既に以下が存在する。

```text
packages/seisai-engine/src/seisai_engine/loss/trend_prior_loss.py
packages/seisai-engine/src/seisai_engine/postprocess/trend_prior_op.py
packages/seisai-pick/src/seisai_pick/trend/trend_fit_strategy.py
```

ただし `seisai_engine.loss.trend_prior_loss.trend_prior_ce()` は function 全体に `@torch.no_grad()` が付いている。これをそのまま Coarse training loss に足すと logits へ勾配が返らない。

この issue では、`packages/*/src/**` を変更せず、site54 OOF scripts 側に学習用の differentiable wrapper を実装し、Issue 003 の criterion hook から base KL loss に合成する。

## 重要な制約

- `packages/seisai-engine/src/**` は変更しない。
- `packages/seisai-pick/src/**` は変更しない。
- 既存 `TrendPriorCELoss` / `trend_prior_ce` は変更しない。
- trend / prior の推定は no-grad でよいが、最終 CE は logits に対して gradient を持つこと。
- baseline / `trend_prior_loss.enabled: false` では挙動を変えない。
- Coarse logits の time axis は 2048 bins なので、valid mask には raw `fb_idx` ではなく `fb_idx_coarse` を使う。

## 現在の repo 前提

Coarse batch には以下が入る。

```text
batch["offsets"]       # (B,H), anchor offsets [m]
batch["dt_sec"]        # effective Coarse dt [s]
batch["fb_idx"]        # raw FB sample index
batch["fb_idx_coarse"] # Coarse time axis sample index, invalid=-1
```

既存 trend helper:

```python
from seisai_pick.score.confidence_from_prob import trace_confidence_from_prob
from seisai_pick.trend._time_pick import _argmax_time_parabolic
from seisai_pick.trend.gaussian_prior_from_trend import gaussian_prior_from_trend
from seisai_pick.trend.trend_fit_strategy import IRLSStrategy, RANSACStrategy
```

## 目的

Coarse training の loss を以下のようにする。

```python
base_loss = base_criterion(pred_for_loss, target, batch)
trend_loss = trend_prior_ce_differentiable(pred_for_loss, batch, cfg)
total_loss = base_loss + weight * trend_loss
```

`pred_for_loss` は Issue 004 の `velocity_mask` 適用後 logits でもよい。`velocity_mask` が無効なら model raw logits。

## 変更対象

主に Issue 003 で追加された runtime files を変更する。

```text
proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py
proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py
```

必要なら以下も更新する。

```text
proc/fbpick/site54/oof/experiments/README.md
proc/fbpick/site54/oof/README.md
```

変更禁止:

```text
packages/seisai-engine/src/**
packages/seisai-pick/src/**
```

## config interface

`site54_oof_experiment.coarse_train_runtime.trend_prior_loss` を読む。

```yaml
site54_oof_experiment:
  coarse_train_runtime:
    enabled: true
    trend_prior_loss:
      enabled: true
      weight: 0.03
      fit_kind: irls
      section_len: 128
      stride: 64
      huber_c: 1.345
      iters: 3
      vmin_m_s: 300.0
      vmax_m_s: 6000.0
      prior_sigma_ms: 20.0
      prior_alpha: 1.0
      prior_conf_gate: 0.5
      prior_log_eps: 1.0e-4
      conf_floor: 0.2
      conf_power: 0.5
      logit_clip: 30.0
      tau: 1.0
      channels: 0
      offsets_key: offsets
      fb_idx_key: fb_idx_coarse
      dt_key: dt_sec
      aux_key: trend_prior
```

### default 値の考え方

- baseline では section が無い、または `enabled: false`。
- `conf_floor`, `conf_power`, `prior_sigma_ms`, `prior_alpha`, `prior_conf_gate`, `prior_log_eps`, `logit_clip`, `channels` は既存 `TrendPriorConfig` の default と揃える。
- `section_len`, `stride`, `huber_c`, `iters` は既存 `IRLSStrategy` default と揃える。
- site54 初期 experiment では、速度範囲は repo 既存 `physical_band` に合わせて `300..6000 m/s` とする。
- `fb_idx_key` は Coarse 用に必ず `fb_idx_coarse` を default とする。

### 1. local differentiable CE を実装する

`trend_prior_ce()` のロジックを参考に、site54 local file に `trend_prior_ce_differentiable()` を実装する。

重要: prior 作成は no-grad でよいが、CE 本体は no-grad にしない。

概念:

```python
def trend_prior_ce_differentiable(logits, batch, cfg):
    # logits: (B,C,H,W)
    # batch relevant tensors are moved to logits.device

    with torch.no_grad():
        # sanitize logits only for fitting
        # prob = softmax(logit_for_fit)
        # confidence, t_sec, trend fit, gaussian prior を作る
        # valid = batch[fb_idx_key] >= 0
        # covered を得る

    # ここから gradient あり
    log_p = torch.log_softmax(logits[:, c] / tau, dim=-1)
    ce_bh = -(prior.detach() * log_p).sum(dim=-1)
    ce = ce_bh[valid & covered].mean()
    return ce
```

### 2. no-grad 範囲

以下は no-grad でよい。

- `trace_confidence_from_prob`
- `_argmax_time_parabolic`
- trend fit strategy call
- `gaussian_prior_from_trend`
- `covered` / `valid` / diagnostics 作成

以下は gradient が必要。

- `log_p = torch.log_softmax(logits[:, c] / tau, dim=-1)`
- `ce_bh = -(prior.detach() * log_p).sum(dim=-1)`
- `ce.mean()`

### 3. fit strategy builder

`fit_kind` は少なくとも以下を support する。

```text
irls
ransac
```

`irls`:

```python
IRLSStrategy(
    section_len=section_len,
    stride=stride,
    huber_c=huber_c,
    iters=iters,
    vmin=vmin_m_s,
    vmax=vmax_m_s,
)
```

`ransac` は既存 `RANSACStrategy` の引数を必要に応じて config から読む。未指定値は class default を使う。

### 4. gating / zero loss

既存 `trend_prior_ce()` は channel がすべて `prior_conf_gate` で skip されると assert する。training 中に assert で落ちると実験しにくいため、site54 local wrapper では以下にする。

- `prior_conf_gate` 未満ならその channel は寄与 0。
- すべての channel が skip された場合は `logits.new_zeros(())` を返す。
- `valid & covered` が空の場合も `logits.new_zeros(())` を返す。

このとき diagnostics に skip reason を残す。

### 5. criterion wrapper への合成

Issue 003 の `Site54CoarseExperimentCriterion` で以下を行う。

```python
base_loss = base_criterion(pred_for_loss, target, batch)
if trend_prior_loss.enabled:
    t_loss, t_aux = trend_prior_ce_differentiable(pred_for_loss, batch, cfg)
    total_loss = base_loss + weight * t_loss
else:
    total_loss = base_loss
```

`weight` は finite かつ `>= 0` に validate する。`weight == 0` は no-op と同じ。

### 6. AMP / dtype

- prior / fit は float32 でよい。
- CE は autocast 下でも `torch.log_softmax` が安定するよう、必要なら logits を float32 に上げる。ただし戻り値 loss は Tensor scalar とし、backward できること。
- non-finite が出た場合は、config と channel を含む明確な error を出す。

### 7. diagnostics

step ごとの詳細保存は不要。aux として以下を持てるようにする。

```text
trend_prior_conf_med_ch0
trend_prior_alpha_eff_ch0
trend_prior_v_trend_ch0
trend_prior_covered_ch0
trend_prior_skipped_ch0
```

runtime summary には最低限以下を保存する。

```json
{
  "trend_prior_loss_enabled": true,
  "trend_prior_loss_weight": 0.03,
  "trend_prior_fit_kind": "irls"
}
```

## 受け入れ条件

- `packages/seisai-engine/src/**` と `packages/seisai-pick/src/**` に変更がない。
- `trend_prior_loss.enabled: false` では baseline no-op。
- `trend_prior_loss.enabled: true` で total loss が `base_loss + weight * trend_loss` になる。
- `trend_prior_ce_differentiable()` の戻り loss に対して `loss.backward()` すると logits に gradient が入る。
- `fb_idx_key` default は `fb_idx_coarse`。
- all gated / no valid の場合に assert で落ちず zero loss を返す。
- `fit_kind: irls` が動く。可能なら `fit_kind: ransac` も smoke する。

## 確認コマンド

```bash
cd /workspace

python -m py_compile \
  proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py \
  proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py

python - <<'PY'
import torch
from pathlib import Path
import sys
sys.path.insert(0, str(Path('proc/fbpick/site54/oof/scripts').resolve()))
from site54_coarse_experiment_runtime import trend_prior_ce_differentiable

B, C, H, W = 2, 1, 16, 128
logits = torch.randn(B, C, H, W, requires_grad=True)
batch = {
    'offsets': torch.linspace(0, 1500, H).repeat(B, 1),
    'dt_sec': torch.full((B,), 0.004),
    'fb_idx_coarse': torch.full((B, H), 20, dtype=torch.long),
}
cfg = {
    'enabled': True,
    'weight': 0.03,
    'fit_kind': 'irls',
    'section_len': 8,
    'stride': 4,
    'huber_c': 1.345,
    'iters': 1,
    'vmin_m_s': 300.0,
    'vmax_m_s': 6000.0,
    'prior_sigma_ms': 20.0,
    'prior_alpha': 1.0,
    'prior_conf_gate': 0.0,
    'prior_log_eps': 1.0e-4,
    'conf_floor': 0.2,
    'conf_power': 0.5,
    'logit_clip': 30.0,
    'tau': 1.0,
    'channels': 0,
    'offsets_key': 'offsets',
    'fb_idx_key': 'fb_idx_coarse',
    'dt_key': 'dt_sec',
    'aux_key': 'trend_prior',
}
loss, aux = trend_prior_ce_differentiable(logits, batch, cfg, return_aux=True)
assert loss.requires_grad
loss.backward()
assert logits.grad is not None
print('ok', float(loss.detach()), sorted(aux)[:5])
PY
```
